lib/fluent/plugin/filter_kubernetes_metadata.rb

```diff
@@ -305,6 +309,17 @@ def configure(conf)
       end
     end
 
+    def create_client()
+      log.debug 'Creating K8S client'
```

Comment:
I wonder if we need to make this function thread safe.?

Reply:
(Unnecessary character building anecdote warning)

So fun fact: Once upon a time, when I was a new developer on the open source scene at my first conference, I got silly stuff from people on twitter because I made a big deal about how Fluent Bit is faster than Fluentd largely because of the language choice and ruby is very slow. 

And it has this thing called a Global Interpreter Lock, which means the multiple threads are never really every actually executing in parallel, its concurrency, not true parallelization, and it just means that when one thread is waiting on the network then another can do stuff, so its not possible for the watcher threads to try to refresh the client at the same time as the main thread.

Reply:
Hmm actually thinking about it some more... may be this isn't completely safe...

Can you help me understand how the watch functionality works- in watch mode, will the client calls in the main file (filter_kubernetes.rb) ever be made? 

I just had the thought that may be if one watch thread or the main thread refreshes the client, that will somehow break the watch on the other watch thread. But if that happens, the other thread would have run into an unauthorized exception anyway... so I think its fine actually. 

I think I just need to do some more testing of this on a real cluster with lots of updates happening over a decent period of time and then I can be confident its safe.

Can you think of any scenarios that this code change will likely break?

Reply:
> I just had the thought that may be if one watch thread or the main thread refreshes the client, that will somehow break the watch on the other watch thread. But if that happens, the other thread would have run into an unauthorized exception anyway... so I think its fine actually.

This was my original thought too.  The watch will more or less fail at the same time so you have a momentary blip in failure that will should converge back to a valid client capable of making calls.